lisa/tools/hibernation_setup.py

```diff
@@ -0,0 +1,66 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+from __future__ import annotations
+
+import re
+from typing import List, Type
+
+from lisa.executable import Tool
+from lisa.util import find_patterns_in_lines
+
+from .dmesg import Dmesg
+from .git import Git
+from .make import Make
+
+
+class HibernationSetup(Tool):
+    repo = "https://github.com/microsoft/hibernation-setup-tool"
+    entry_pattern = re.compile(r"^(.*hibernation entry.*)$", re.MULTILINE)
+    exit_pattern = re.compile(r"^(.*hibernation exit.*)$", re.MULTILINE)
+
+    @property
+    def command(self) -> str:
+        return "hibernation-setup-tool"
+
+    @property
+    def dependencies(self) -> List[Type[Tool]]:
+        return [Git, Make]
+
+    @property
+    def can_install(self) -> bool:
+        return True
+
+    def _install(self) -> bool:
+        tool_path = self.get_tool_path()
+        git = self.node.tools[Git]
+        git.clone(self.repo, tool_path)
+        code_path = tool_path.joinpath("hibernation-setup-tool")
+        make = self.node.tools[Make]
+        make.make_install(code_path)
+        return self._check_exists()
+
+    def start(self) -> None:
+        self.run(
+            sudo=True,
+            expected_exit_code=0,
+            expected_exit_code_failure_message="fail to start",
+        )
+
+    def check_entry(self) -> int:
+        dmesg = self.node.tools[Dmesg]
+        dmesg_output = dmesg.get_output(force_run=True)
+        matched_lines = find_patterns_in_lines(dmesg_output, [self.entry_pattern])
+        if not matched_lines:
+            return 0
+        return len(matched_lines[0])
+
+    def check_exit(self) -> int:
```

Comment:
check_entry and check_exit are pretty much same with except the matched pattern. Maybe we can add a _check function(self, pattern) and call it from check_entry and check_exit

Reply:
fixed